File: src/main/java/edu/cornell/cs/apl/viaduct/imp/ast/JumpLabel.java
package edu.cornell.cs.apl.viaduct.imp.ast;

import com.google.auto.value.AutoValue;

/** Target labels for unstructured control statements like {@code continue} and {@code break}. */
@AutoValue
public abstract class JumpLabel extends Located implements Name {
  public static Builder builder() {
    return new AutoValue_JumpLabel.Builder();
  }

  public abstract Builder toBuilder();

  @Override
  public final String getNameCategory() {
    return "label";
  }

  @Override
  public final String toString() {
    return getName();
  }

  @AutoValue.Builder
  public abstract static class Builder extends Located.Builder<Builder> {
    public abstract Builder setName(String name);

    public abstract JumpLabel build();
  }
}

File: src/main/kotlin/edu/cornell/cs/apl/viaduct/syntax/JumpLabel.kt
package edu.cornell.cs.apl.viaduct.syntax

import edu.cornell.cs.apl.prettyprinting.AnsiBaseColor
import edu.cornell.cs.apl.prettyprinting.AnsiColor
import edu.cornell.cs.apl.prettyprinting.Document
import edu.cornell.cs.apl.prettyprinting.NormalColor
import edu.cornell.cs.apl.prettyprinting.Style
import edu.cornell.cs.apl.prettyprinting.styled

/** The target label for unstructured control statements like `continue` and `break`. */
data class JumpLabel(override val name: String) : Name {
    override val nameCategory: String
        get() = "label"

    override val asDocument: Document
        get() = Document(name).styled(JumpLabelStyle)
}

/** The display style of [JumpLabel]s. */
object JumpLabelStyle : Style {
    override val foregroundColor: AnsiColor
        get() = NormalColor(AnsiBaseColor.BLUE)
}
